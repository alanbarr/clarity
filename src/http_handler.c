/*******************************************************************************
* Copyright (c) 2014, Alan Barr
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
*
* * Redistributions of source code must retain the above copyright notice, this
*   list of conditions and the following disclaimer.
*
* * Redistributions in binary form must reproduce the above copyright notice,
*   this list of conditions and the following disclaimer in the documentation
*   and/or other materials provided with the distribution.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*******************************************************************************/

#include <strings.h>
#include "clarity_int.h"
#include "clarity_api.h"
#include "socket.h"

/* TODO use to cut short processing */
#if 0
int32_t httpHandlerCheckResouceMethod(clarityHttpRequestInformation * info)
{
    (void)info;
    return true;
}
#endif

static int32_t getMethod(clarityResourceInformation * resInfo,
                         clarityHttpMethodType method, 
                         clarityHttpMethodInformation ** methodInfo)
{
    uint8_t methIndex;

    for (methIndex=0; methIndex<CLARITY_MAX_METHODS; methIndex++)
    {
        if (resInfo->methods[methIndex].type == method)
        {
            *methodInfo=&resInfo->methods[methIndex];
            return true;
        }
    }

    return false;
}

static int32_t getResource(clarityHttpServerInformation * con, clarityBuf * resBuf,
                           clarityResourceInformation ** resInfo)
{
    uint8_t resIndex;
    for (resIndex=0; resIndex<CLARITY_MAX_RESOURCES; resIndex++)
    {
        if(strncasecmp(resBuf->data, con->resources[resIndex].name,
                       resBuf->size) == 0)
        {
            *resInfo=&con->resources[resIndex];
            return true;
        }
    }
    return false;
}


int32_t httpHandle(clarityHttpRequestInformation * info,
                   clarityHttpServerInformation * control, void * user)
{
    clarityResourceInformation * resInfo;
    clarityHttpMethodInformation * methInfo;

    if (true != getResource(control, &info->resource, &resInfo))
    {
        CLAR_PRINT_ERROR();
    }

    else if (true != getMethod(resInfo, info->type, &methInfo))
    {
        CLAR_PRINT_ERROR();
    }

    else
    {
        if (methInfo->callback != NULL)
        {
            methInfo->callback(info, user);
        }
        else
        {
            CLAR_PRINT_ERROR();
        }

        return true;
    }

    return false;
}


const char * httpRequestProcess(clarityHttpServerInformation * control,
                                clarityHttpRequestInformation * http,
                                clarityConnectionInformation * connection,
                                const char * data, uint16_t size)
{
    const char * rtnd = NULL;
    
    if (NULL == (rtnd = httpParseRequest(http, data, size)))
    {
        CLAR_PRINT_ERROR();
        return NULL;
    }

    else if (httpHandle(http, control, connection) != true)
    {
        CLAR_PRINT_ERROR();
        return NULL;
    }

    else 
    {
        return rtnd;
    }

    return NULL;
}



