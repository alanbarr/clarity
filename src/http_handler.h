#include <stdint.h>
#include <stdbool.h>

#define DEVICE_NAME_SIZE            15
#define RESOURCE_NAME_SIZE          20

#define RECEIVE_BUFFER_SIZE         1024
#define RECEIVE_BUFFER_LAST_INDEX   (RECEIVE_BUFFER_SIZE - 1)

#define TRANSMIT_BUFFER_SIZE        1024

#define MAX_RESOURCES               4   /* Per control */
#define MAX_METHODS                 4   /* Per resource */

#define HTTP_EOL_STR                "\r\n"
#define HTTP_EOL_LEN                2

#define HTTP_VERSION_LEN            8   /* "HTTP/X.X" */
                                    /* "GET / " VERSION \r\n */
#define HTTP_START_LINE_MIN_LEN     (6 + HTTP_VERSION_LEN) 
#define HTTP_SPACE_LEN              1   /* ' ' */

#define ASCII_DIGIT_OFFSET          48

#define HTTP_RESOURCE_START_CHAR    '/'

#define PRINT_MESSAGES              FALSE

#if PRINT_MESSAGES == TRUE  
    #define PRINT(STR, ...) \
        printf(STR, __VA_ARGS__)
    
    #define PRINT_LINE(STR) \
        printf("(%s %s:%d) " STR "\r\n", __FUNCTION__, __FILE__, __LINE__)
    
    #define PRINT_LINE_ARGS(STR, ...) \
        printf("(%s %s:%d): " STR "\r\n", __FUNCTION__, __FILE__, __LINE__, __VA_ARGS__)
#else
    #define PRINT(...) 
    #define PRINT_LINE(...) 
    #define PRINT_LINE_ARGS(...) 
#endif

#define METHODS     \
    METHOD(GET)     \
    METHOD(HEAD)    \
    METHOD(PUT)     \
    METHOD(POST)    \
    METHOD(TRACE)   \
    METHOD(OPTIONS) \
    METHOD(DELETE)  \

#ifdef METHOD
    #undef METHOD
#endif 
#define METHOD(M)    M,

typedef enum {
    METHODS
    METHOD_INVALID
} methodType;

typedef struct {
    const char * data; 
    uint16_t size;
} buf;

typedef struct {
    uint8_t major;
    uint8_t minor;
} httpVersion;

typedef struct {
    methodType type;
    buf resource;
    httpVersion version;


} httpParser;


typedef struct {
    char buffer[RECEIVE_BUFFER_SIZE];
    uint16_t start;
    uint16_t end;
    uint16_t taken;
} readBuffer;


typedef enum {
    HEADER_TYPE_NONE,
    CONTENT_TYPE
} headerType;


typedef uint32_t (*methodCallback)(const char * body, const char * data,
                                   uint16_t size, void * user);

typedef struct {
    methodType type;
    methodCallback callback;
} methodInformation;


typedef struct {
    char name[RESOURCE_NAME_SIZE];
    methodType methodsMask; /* Mask of all supported methods */
    methodInformation methods[MAX_METHODS];
} resourceInformation;


typedef struct {
    resourceInformation * resource;
    methodInformation * method;
    headerType nextHeaderValue;
    bool completed;
} parsingInformation;


typedef struct { 
    char deviceName[DEVICE_NAME_SIZE]; 
    resourceInformation resources[MAX_RESOURCES];
    parsingInformation parsed;
} controlInformation;

const char * parseHttp(httpParser * par, const char * data, const uint16_t size,
               void * user);

void registerControl(controlInformation * control);



