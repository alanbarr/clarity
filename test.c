#include <string.h>
#include <stdio.h>
#include "http_parser.h"
#include "http_handler.h"

char * tst_get =
"GET /robots.txt HTTP/1.0\r\n"
"Host: www.joes-hardware.com\r\n"
"User-Agent: Slurp/2.0\r\n"
"Date: Wed Oct 3 20:22:48 EST 2001\r\n"
"\r\n";

char * tst_put =
"POST /path/script.cgi HTTP/1.0\r\n"
"From: frog@jmarshall.com\r\n"
"User-Agent: HTTPTool/1.0\r\n"
"Content-Type: application/x-www-form-urlencoded\r\n"
"Content-Length: 32\r\n"
"\r\n"
"home=Cosby&favorite+flavor=flies\r\n";

char * tst_root =
"GET / HTTP/1.0\r\n"
"\r\n";



uint32_t mcb(const char * body, const char * data, const uint16_t size, void * user);
uint32_t getRoot(const char * body, const char * data, const uint16_t size, void * user);

controlInformation control =  /* TODO only one? */
{
    "Test_Control",     /* Device Name */
    /* Resources */
    {

        /* Resource 1 */
        {
            "/",
            GET,
            /* Methods */
            {
                {
                    GET,
                    getRoot
                }
            }
        },
        /* Resource 2 */
        {
            "/path/script.cgi", /* Resource Name */
            POST,               /* Methods Supported Mask */
            /* Methods */
            {
                /* Method 1 */
                {
                    POST,       /* Method Type */
                    mcb         /* Callback */
                }
            }
        },
        /* Resource 3 */
        {
            "/robots.txt",   /* Resource Name */
            GET | PUT,       /* Methods Supported Mask */
            /* Methods */
            {
                /* Method 1 */
                {
                    GET,            /* Method Type */
                    mcb             /* Callback */
                },
                /* Method 2 */
                {
                    PUT,            /* Method Type */
                    mcb             /* Callback */
                }
            },
        }

    },
    {
    }
};



uint32_t mcb(const char * body, const char * data, const uint16_t size, void * user)
{
    PRINT_LINE("in MCB CB!!", NULL);
    return 0;
}   

uint32_t getRoot(const char * body, const char * data, const uint16_t size, void * user)
{

static char * html = 
        "<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">"
        "<html><head>"
        "<title>403 Forbidden</title>"
        "</head><body>"
        "<h1>Forbidden</h1>"
        "<p>You don't have permission to access /"
        "on this server.</p>"
        "</body></html>";

    PRINT_LINE("IN GET ROOT", NULL);

    return 0;
}   



int main(void) 
{
    registerControl(&control);
#if 0
    parseHttp(tst_get, strlen(tst_get), NULL);
#endif
#if 0
    parseHttp(tst_put, strlen(tst_put), NULL);
#endif
#if 1
    parseHttp(tst_root, strlen(tst_root), NULL);
#endif
}

