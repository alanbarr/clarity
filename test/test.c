#include <string.h>
#include <stdio.h>

#if 0
#include "http_parser.h"
#endif
#include "http_handler.h"

#define TEST_PRINT(S) \
    printf("TEST (%s:%d) " S "\n", __FILE__, __LINE__)

#define TEST_PRINT_ARG(S, ...) \
    printf("TEST (%s:%d) " S "\n", __FILE__, __LINE__, __VA_ARGS__)

#define ERROR_CHECK(BOOL, STR) \
    test_check(&testData, BOOL, STR, __FILE__, __LINE__)

char * tst_get =
"GET /robots.txt HTTP/1.0\r\n"
"Host: www.joes-hardware.com\r\n"
"User-Agent: Slurp/2.0\r\n"
"Date: Wed Oct 3 20:22:48 EST 2001\r\n"
"\r\n";

char * tst_put =
"POST /path/script.cgi HTTP/1.0\r\n"
"From: frog@jmarshall.com\r\n"
"User-Agent: HTTPTool/1.0\r\n"
"Content-Type: application/x-www-form-urlencoded\r\n"
"Content-Length: 32\r\n"
"\r\n"
"home=Cosby&favorite+flavor=flies\r\n";

char * tst_root =
"GET / HTTP/1.0\r\n"
"\r\n";
 
typedef struct {
    uint32_t checks;
    uint32_t passes;
    uint32_t fails;
} testInformation;

static testInformation testData;

uint32_t mcb(const char * body, const char * data, const uint16_t size, void * user);
uint32_t getRoot(const char * body, const char * data, const uint16_t size, void * user);

controlInformation control =  /* TODO only one? */
{
    "Test_Control",     /* Device Name */
    /* Resources */
    {
        /* Resource 1 */
        {
            "/",
            GET,
            /* Methods */
            {
                {
                    GET,
                    getRoot
                }
            }
        },
        /* Resource 2 */
        {
            "/path/script.cgi", /* Resource Name */
            POST,               /* Methods Supported Mask */
            /* Methods */
            {
                /* Method 1 */
                {
                    POST,       /* Method Type */
                    mcb         /* Callback */
                }
            }
        },
        /* Resource 3 */
        {
            "/robots.txt",   /* Resource Name */
            GET | PUT,       /* Methods Supported Mask */
            /* Methods */
            {
                /* Method 1 */
                {
                    GET,            /* Method Type */
                    mcb             /* Callback */
                },
                /* Method 2 */
                {
                    PUT,            /* Method Type */
                    mcb             /* Callback */
                }
            },
        }

    },
    {
    }
};



uint32_t mcb(const char * body, const char * data, const uint16_t size, void * user)
{
    TEST_PRINT("in MCB CB!!");
    return 0;
}   

uint32_t getRoot(const char * body, const char * data, const uint16_t size, void * user)
{

#if 0
    static char * html = 
        "<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">"
        "<html><head>"
        "<title>403 Forbidden</title>"
        "</head><body>"
        "<h1>Forbidden</h1>"
        "<p>You don't have permission to access /"
        "on this server.</p>"
        "</body></html>";
#endif


    TEST_PRINT("IN GET ROOT");

    return 0;
}   


static void test_check(testInformation * test, 
                       bool error, char * string,
                       char * file, int line)
{
    test->checks++;

    if (error)
    {
        printf("FAILED (%s:%d): %s\n", file,line,string);
        test->fails++;
    }
    else
    {
        test->passes++;
    }
}

static void test_print(testInformation * test)
{
    printf("--Test Information--\n");
    printf("Checks: %d\n", test->checks);
    printf("Passes: %d\n", test->passes);
    printf("Fails: %d\n", test->fails);
}

static int test_startLinesValid(void)
{
    int i;
    httpParser par;
    const char * rtn = NULL;

    static const char * startLinesValid[] = {
        "GET / HTTP/1.1 \r\n"
         "\r\n",
        "HEAD /robots.txt HTTP/0.9 \r\n"
         "\r\n",
        "PUT /one/two/three/four HTTP/2.0\r\n"
         "\r\n",
        "POST / HTTP/0.9 \r\n"
         "\r\n",
        "TRACE / HTTP/0.9 \r\n"
         "\r\n",
        "OPTIONS /onetwothreefourfivesix HTTP/0.9 \r\n"
         "\r\n",
        "DELETE /robots.txt HTTP/0.9 \r\n"
         "\r\n"
    };

    static httpParser startLineDataValid[7];

    memset(&startLineDataValid, 0, sizeof(startLineDataValid));

    startLineDataValid[0].type = GET; 
    startLineDataValid[0].resource.size = 1; 
    startLineDataValid[0].resource.data = startLinesValid[0] + 4;  
    startLineDataValid[0].version.major = 1; 
    startLineDataValid[0].version.minor = 1; 

    startLineDataValid[1].type = HEAD; 
    startLineDataValid[1].resource.size = 11; 
    startLineDataValid[1].resource.data = startLinesValid[1] + 5;  
    startLineDataValid[1].version.major = 0; 
    startLineDataValid[1].version.minor = 9; 

    startLineDataValid[2].type = PUT; 
    startLineDataValid[2].resource.size = 19; 
    startLineDataValid[2].resource.data = startLinesValid[2] + 4;  
    startLineDataValid[2].version.major = 2; 
    startLineDataValid[2].version.minor = 0; 

    startLineDataValid[3].type = POST; 
    startLineDataValid[3].resource.size = 1; 
    startLineDataValid[3].resource.data = startLinesValid[3] + 5;  
    startLineDataValid[3].version.major = 0; 
    startLineDataValid[3].version.minor = 9; 

    startLineDataValid[4].type = TRACE; 
    startLineDataValid[4].resource.size = 1; 
    startLineDataValid[4].resource.data = startLinesValid[4] + 6;  
    startLineDataValid[4].version.major = 0; 
    startLineDataValid[4].version.minor = 9; 
 
    startLineDataValid[5].type = OPTIONS; 
    startLineDataValid[5].resource.size = 23; 
    startLineDataValid[5].resource.data = startLinesValid[5] + 8;  
    startLineDataValid[5].version.major = 0; 
    startLineDataValid[5].version.minor = 9; 

    startLineDataValid[6].type = DELETE; 
    startLineDataValid[6].resource.size = 11; 
    startLineDataValid[6].resource.data = startLinesValid[6] + 7;  
    startLineDataValid[6].version.major = 0; 
    startLineDataValid[6].version.minor = 9; 

    for(i=0;i<7;i++)
    {
        TEST_PRINT_ARG("loop iternation %d.", i);
        memset(&par, 0, sizeof(par));

        rtn = parseHttp(&par, startLinesValid[i], strlen(startLinesValid[i]),
                        NULL);

        ERROR_CHECK(rtn == NULL,"RTN was NULL");
     
        ERROR_CHECK(startLineDataValid[i].type != par.type,
                    "Type not correct.");
        ERROR_CHECK(startLineDataValid[i].resource.size != par.resource.size,
                    "Resource size not correct.");
        ERROR_CHECK(startLineDataValid[i].resource.data != par.resource.data,
                    "Resource data not correct.");  
        ERROR_CHECK(startLineDataValid[i].version.major != par.version.major,
                    "Version major not correct."); 
        ERROR_CHECK(startLineDataValid[i].version.minor != par.version.minor,
                    "Version minor not correct."); 
    }
    return 1;
}

static int test_startLinesInvalid(void)
{
    int i;
    httpParser par;
    const char * rtn = NULL;
    
    static const char * startLinesInvalid[] = {
    /* No space between method and resource */
    "GET/ HTTP/1.1 \r\n",    
    /* No CRLF */
    "HEAD /robots.txt HTTP/0.9",
    /* No resource */
    "PUT HTTP/2.0\r\n",
    /* No sapce between resource and version */
    "POST /HTTP/0.9\r\n",
    /* Unknown method */
    "TR / HTTP/0.9 \r\n",
    /* Unknown method */
    "OPTION /onetwothreefourfivesix HTTP/0.9 \r\n",
    /* Newline missing */
    "DELETE /robots.txt HTTP/0.9 \r"
};



    
    for (i=0;i<7;i++)
    {
        memset(&par, 0, sizeof(par));
        rtn = parseHttp(&par, startLinesInvalid[i], strlen(startLinesInvalid[i]),
                        NULL);
        ERROR_CHECK(rtn != NULL,"RTN was not NULL");
    }
    return 1;
}

static int test_headersValid(void)
{
    int index;
    int hindex;
    httpParser par;
    const char * rtn = NULL;
    static httpParser headerData[3];

    static const char * headersValid[] = {
    "DELETE /robots.txt HTTP/0.9 \r\n"
    "HEADER FIELD: HEADER VALUE\r\n"
    "\r\n",
    "DELETE /robots.txt HTTP/0.9 \r\n"
    "HEADER FIELD: HEADER VALUE\r\n"
    "HEADER FIELD: HEADER VALUE\r\n"
    "\r\n",

    "GET / HTTP/2.0 \r\n"
    "HEADER FIELD: HEADER VALUE\r\n"
    "HEADER FIELD: HEADER VALUE\r\n"
    "HEADER FIELD: HEADER VALUE\r\n"
    "HEADER FIELD: HEADER VALUE\r\n"
    "HEADER FIELD: HEADER VALUE\r\n"
    "HEADER FIELD: HEADER VALUE\r\n"
    "\r\n",

    "PUT /robots.txt HTTP/0.9 \r\n"
    "Content-Type: text/plain\r\n"
    "Content-Length: 1\r\n"
    "\r\n"
    "1",
    };

    memset(&headerData, 0, sizeof(headerData));

    headerData[0].type = DELETE; 
    headerData[0].resource.size = 11; 
    headerData[0].resource.data = headersValid[0] + 7;  
    headerData[0].version.major = 0; 
    headerData[0].version.minor = 9; 
    headerData[0].headers[0].field.data = headersValid[0] + 30; 
    headerData[0].headers[0].field.size = 12; 
    headerData[0].headers[0].value.data = headersValid[0] + 44; 
    headerData[0].headers[0].value.size = 12; 

    headerData[1].type = DELETE; 
    headerData[1].resource.size = 11; 
    headerData[1].resource.data = headersValid[1] + 7;  
    headerData[1].version.major = 0; 
    headerData[1].version.minor = 9; 
    headerData[1].headers[0].field.data = headersValid[1] + 30; 
    headerData[1].headers[0].field.size = 12; 
    headerData[1].headers[0].value.data = headersValid[1] + 30 + (14); 
    headerData[1].headers[0].value.size = 12; 
    headerData[1].headers[1].field.data = headersValid[1] + 30 + (14*2); 
    headerData[1].headers[1].field.size = 12; 
    headerData[1].headers[1].value.data = headersValid[1] + 30 + (14*3); 
    headerData[1].headers[1].value.size = 12; 
 
    headerData[2].type = GET; 
    headerData[2].resource.size = 1; 
    headerData[2].resource.data = headersValid[2] + 4;  
    headerData[2].version.major = 2; 
    headerData[2].version.minor = 0; 
    headerData[2].headers[0].field.data = headersValid[2] + 17; 
    headerData[2].headers[0].field.size = 12; 
    headerData[2].headers[0].value.data = headersValid[2] + 17 + (14); 
    headerData[2].headers[0].value.size = 12; 
    headerData[2].headers[1].field.data = headersValid[2] + 17 + (14*2); 
    headerData[2].headers[1].field.size = 12; 
    headerData[2].headers[1].value.data = headersValid[2] + 17 + (14*3); 
    headerData[2].headers[1].value.size = 12; 
    headerData[2].headers[2].field.data = headersValid[2] + 17 + (14*4); 
    headerData[2].headers[2].field.size = 12; 
    headerData[2].headers[2].value.data = headersValid[2] + 17 + (14*5);
    headerData[2].headers[2].value.size = 12; 
    headerData[2].headers[3].field.data = headersValid[2] + 17 + (14*6);
    headerData[2].headers[3].field.size = 12; 
    headerData[2].headers[3].value.data = headersValid[2] + 17 + (14*7);
    headerData[2].headers[3].value.size = 12; 
    headerData[2].headers[4].field.data = headersValid[2] + 17 + (14*8);
    headerData[2].headers[4].field.size = 12; 
    headerData[2].headers[4].value.data = headersValid[2] + 17 + (14*9);
    headerData[2].headers[4].value.size = 12; 
    headerData[2].headers[5].field.data = headersValid[2] + 17 + (14*10);
    headerData[2].headers[5].field.size = 12; 
    headerData[2].headers[5].value.data = headersValid[2] + 17 + (14*11);
    headerData[2].headers[5].value.size = 12; 

    headerData[3].type = PUT; 
    headerData[3].resource.size = 11; 
    headerData[3].resource.data = headersValid[3] + 7;  
    headerData[3].version.major = 0; 
    headerData[3].version.minor = 9; 
    headerData[3].content.data = headersValid[3] + 41;
    headerData[3].content.size = 1;

    for (index = 0; index<3; index++)
    {
        TEST_PRINT_ARG("Checking index: %d.", index);

        memset(&par,0,sizeof(par));
        rtn = parseHttp(&par, headersValid[index], strlen(headersValid[index]),
                        NULL);
        ERROR_CHECK(rtn == NULL,"RTN was NULL");  
        ERROR_CHECK(headerData[index].type != par.type,
                    "Type not correct.");
        ERROR_CHECK(headerData[index].resource.size != par.resource.size,
                    "Resource size not correct.");
        ERROR_CHECK(headerData[index].resource.data != par.resource.data,
                    "Resource data not correct.");  
        ERROR_CHECK(headerData[index].version.major != par.version.major,
                    "Version major not correct."); 
        ERROR_CHECK(headerData[index].version.minor != par.version.minor,
                    "Version minor not correct."); 
        ERROR_CHECK(headerData[index].content.size != par.content.size,
                    "content size not correct.");
        ERROR_CHECK(headerData[index].content.data != par.content.data,
                    "content data not correct.");  

        for (hindex = 0; hindex< MAX_HEADERS; hindex++)
        {
            TEST_PRINT_ARG("Checking header index: %d.", hindex);

            ERROR_CHECK(par.headers[hindex].field.data   != 
                        headerData[index].headers[hindex].field.data,
                        "Header field data not as expected");
            ERROR_CHECK(par.headers[hindex].field.size !=
                        headerData[index].headers[hindex].field.size,
                        "Header field size not as expected");
            ERROR_CHECK(par.headers[hindex].value.data   != 
                        headerData[index].headers[hindex].value.data,
                        "Header value data not as expected");
            ERROR_CHECK(par.headers[hindex].value.size !=
                        headerData[index].headers[hindex].value.size,
                        "Header value size not as expected");
        }
    }
    return 1;
}


static int test_headersInvalid(void)
{
    int index;
    int hindex;
    httpParser par;
    const char * rtn = NULL;
    static httpParser headerData[2];

    static const char * headersInvalid[] = {
    "GET / HTTP/1.0 \r\n"
    "HEADER FIELD:\r\n"
    "\r\n",
    "GET / HTTP/1.0 \r\n"
    "HEADER FIELD\r\n"
    "\r\n",
    "GET / HTTP/1.0 \r\n"
    "HEADER FIELD: HEADER VALUE\r\n"
    "HEADER FIELD\r\n"
    "\r\n",
    };

    memset(&headerData, 0, sizeof(headerData));

    headerData[0].type = GET; 
    headerData[0].resource.size = 1; 
    headerData[0].resource.data = headersInvalid[0] + 4;  
    headerData[0].version.major = 1; 
    headerData[0].version.minor = 0; 

    headerData[1].type = GET; 
    headerData[1].resource.size = 1; 
    headerData[1].resource.data = headersInvalid[1] + 4;  
    headerData[1].version.major = 1; 
    headerData[1].version.minor = 0; 

    headerData[2].type = GET; 
    headerData[2].resource.size = 1; 
    headerData[2].resource.data = headersInvalid[2] + 4;  
    headerData[2].version.major = 1; 
    headerData[2].version.minor = 0; 
    headerData[2].headers[0].field.data = headersInvalid[2] + 17; 
    headerData[2].headers[0].field.size = 12; 
    headerData[2].headers[0].value.data = headersInvalid[2] + 17+14; 
    headerData[2].headers[0].value.size = 12; 


    for (index = 0; index<3; index++)
    {
        TEST_PRINT_ARG("Checking index: %d.", index);

        memset(&par,0,sizeof(par));
        rtn = parseHttp(&par, headersInvalid[index], strlen(headersInvalid[index]),
                        NULL);

        ERROR_CHECK(rtn != NULL,"RTN was not NULL");

        ERROR_CHECK(headerData[index].type != par.type,
                    "Type not correct.");
        ERROR_CHECK(headerData[index].resource.size != par.resource.size,
                    "Resource size not correct.");
        ERROR_CHECK(headerData[index].resource.data != par.resource.data,
                    "Resource data not correct.");  
        ERROR_CHECK(headerData[index].version.major != par.version.major,
                    "Version major not correct."); 
        ERROR_CHECK(headerData[index].version.minor != par.version.minor,
                    "Version minor not correct."); 

        for (hindex = 0; hindex< MAX_HEADERS; hindex++)
        {
            TEST_PRINT_ARG("Checking header index: %d.", hindex);

            ERROR_CHECK(par.headers[hindex].field.data   != 
                        headerData[index].headers[hindex].field.data,
                        "Header field data not as expected");
            ERROR_CHECK(par.headers[hindex].field.size !=
                        headerData[index].headers[hindex].field.size,
                        "Header field size not as expected");
            ERROR_CHECK(par.headers[hindex].value.data   != 
                        headerData[index].headers[hindex].value.data,
                        "Header value data not as expected");
            ERROR_CHECK(par.headers[hindex].value.size !=
                        headerData[index].headers[hindex].value.size,
                        "Header value size not as expected");
        }
    }
    return 1;
}


static int test_bodyValid(void)
{
    httpParser par;
    int index;
    const char * rtn;

    static const char * bodiesValid[] = {   
        "PUT / HTTP/1.0\r\n"                
        "Content-Type: text/plain\r\n"      
        "Content-Length: 4\r\n"             
        "\r\n"                              
        "BODY",                             
 
        "POST / HTTP/1.0 \r\n"              
        "HEADER FIELD: HEADER VALUE\r\n"    
        "Content-Type: text/plain\r\n"      
        "Content-Length: 15\r\n"            
        "HEADER FIELD: HEADER VALUE\r\n"    
        "\r\n"                              
        "BODY12345678910"                   
    };

    static httpParser bodyData[2];

    bodyData[0].type = GET; 
    bodyData[0].resource.size = 1; 
    bodyData[0].resource.data = bodiesValid[0] + 4;  
    bodyData[0].version.major = 1; 
    bodyData[0].version.minor = 0; 
    bodyData[0].content.data = bodiesValid[0] + 30;
    bodyData[0].content.size = 10;
    bodyData[0].body.data = bodiesValid[0] + 63; 
    bodyData[0].body.size = 4; 

    bodyData[1].type = POST; 
    bodyData[1].resource.size = 1; 
    bodyData[1].resource.data = bodiesValid[0] + 5;  
    bodyData[1].version.major = 1; 
    bodyData[1].version.minor = 0; 
    bodyData[1].content.data = bodiesValid[1] + 60;
    bodyData[1].content.size = 10;
    bodyData[1].body.data = bodiesValid[1] + 122; 
    bodyData[1].body.size = 15; 

    for (index = 0; index < 2; index++)
    {
        TEST_PRINT_ARG("Checking index: %d.", index);

        memset(&par,0,sizeof(par));
        rtn = parseHttp(&par, bodiesValid[index], strlen(bodiesValid[index]),
                        NULL);
        ERROR_CHECK(rtn == NULL,"RTN was NULL");
        ERROR_CHECK(par.body.data != bodyData[index].body.data,
                    "Body Data not as expected.");
        ERROR_CHECK(par.body.size != bodyData[index].body.size,
                    "Body Size not as expected.");
        ERROR_CHECK(par.content.size != bodyData[index].content.size,
                    "Content size not as expected.");
        ERROR_CHECK(par.content.data != bodyData[index].content.data,
                    "Content data not as expected.");
    }

    return 0;
}

int main(void) 
{
#if 0
    httpParser par;
#endif 
    registerControl(&control);
#if 0
    parseHttp(tst_get, strlen(tst_get), NULL);
#endif
#if 0
    parseHttp(&par, tst_put, strlen(tst_put), NULL);
#endif
#if 0
    parseHttp(&par, tst_root, strlen(tst_root), NULL);
#endif

    test_startLinesValid();
    test_startLinesInvalid();
    test_headersValid();
    test_headersInvalid();
    test_bodyValid();
    test_print(&testData);
    return 0;
}

