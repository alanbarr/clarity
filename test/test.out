TEST (test.c:452) Checking index: 0.
(getNextLine ../src/http_handler.c:283) Should be returning success.
(parseHttpHeaders ../src/http_handler.c:447): IN parseHttpHeaders.
(parseHttpHeaders ../src/http_handler.c:518): ALAN DEBUG. Going to try to grab next line.dataStart: 0x402951 size: 17 eol: 0x40295e dataEnd 0x402962
FAILED (test.c:458): RTN was not NULL
TEST (test.c:473) Checking header index: 0.
FAILED (test.c:477): Header field data not as expected
FAILED (test.c:480): Header field size not as expected
FAILED (test.c:483): Header value data not as expected
FAILED (test.c:486): Header value size not as expected
TEST (test.c:473) Checking header index: 1.
TEST (test.c:473) Checking header index: 2.
TEST (test.c:473) Checking header index: 3.
TEST (test.c:473) Checking header index: 4.
TEST (test.c:473) Checking header index: 5.
TEST (test.c:473) Checking header index: 6.
TEST (test.c:473) Checking header index: 7.
TEST (test.c:473) Checking header index: 8.
TEST (test.c:473) Checking header index: 9.
TEST (test.c:452) Checking index: 1.
(getNextLine ../src/http_handler.c:283) Should be returning success.
(parseHttpHeaders ../src/http_handler.c:447): IN parseHttpHeaders.
(parseHttpHeaders ../src/http_handler.c:484) header field length was zero.
(parseHttp ../src/http_handler.c:587) Parse Headers Failed.
TEST (test.c:473) Checking header index: 0.
TEST (test.c:473) Checking header index: 1.
TEST (test.c:473) Checking header index: 2.
TEST (test.c:473) Checking header index: 3.
TEST (test.c:473) Checking header index: 4.
TEST (test.c:473) Checking header index: 5.
TEST (test.c:473) Checking header index: 6.
TEST (test.c:473) Checking header index: 7.
TEST (test.c:473) Checking header index: 8.
TEST (test.c:473) Checking header index: 9.
TEST (test.c:452) Checking index: 2.
(getNextLine ../src/http_handler.c:283) Should be returning success.
(parseHttpHeaders ../src/http_handler.c:447): IN parseHttpHeaders.
(parseHttpHeaders ../src/http_handler.c:518): ALAN DEBUG. Going to try to grab next line.dataStart: 0x4029a1 size: 44 eol: 0x4029bb dataEnd 0x4029cd
(getNextLine ../src/http_handler.c:283) Should be returning success.
(parseHttpHeaders ../src/http_handler.c:484) header field length was zero.
(parseHttp ../src/http_handler.c:587) Parse Headers Failed.
TEST (test.c:473) Checking header index: 0.
TEST (test.c:473) Checking header index: 1.
TEST (test.c:473) Checking header index: 2.
TEST (test.c:473) Checking header index: 3.
TEST (test.c:473) Checking header index: 4.
TEST (test.c:473) Checking header index: 5.
TEST (test.c:473) Checking header index: 6.
TEST (test.c:473) Checking header index: 7.
TEST (test.c:473) Checking header index: 8.
TEST (test.c:473) Checking header index: 9.
--Test Information--
Checks: 138
Passes: 133
Fails: 5
